### ISSUE TITLE
Audio Device for Linux VMs

### ISSUE THREAD

**Issue by adit-s** [cirruslabs/tart#795]
Hello. I came across tart.run a few days ago and just started to experiment with it. So far, it works very well and the performance is so good, it feels almost like bare metal. 
I've gone through the docs and github issues and discussions, but couldn't figure it out so thought I'd ask here: 
Is there a way to add an audio device to a linux VM (either from one of the pre-built images (like ghcr.io/cirruslabs/ubuntu:latest) or when creating from scratch (https://tart.run/integrations/vm-management/)?

Basically, I'd like to try and run a full desktop with sound. Everything else seems to work great following the standard steps except for the audio.

I'm not familiar with the packer plugin and not really sure if it's necessary for this request.

Thanks in advance.

Environment:
tart version 2.9.0
Macbook Pro M1 with Sonoma 14.4.1

**Comment by fkorotkov** [comment]
@adit-s which distribution do you use? Tart provides virtio audio devices.

**Comment by adit-s** [comment]
@fkorotkov I've tried with two:

I started with the pre-built image (not sure if that's the right terminology) - ghcr.io/cirruslabs/ubuntu:latest (which is Ubuntu 22.04)
It didn't work so I tried with a manual installation using Ubuntu 23.10. 

Hovering over the speaker it says "no output or input devices found."

Screenshots from both are attached.
![Screenshot 2024-04-22 at 11 31 01 PM](https://github.com/cirruslabs/tart/assets/8761424/d77ddc37-9a55-4601-97e4-9c91e6ee4b71)
![Screenshot 2024-04-22 at 11 35 27 PM](https://github.com/cirruslabs/tart/assets/8761424/4ee6f3a8-0c67-46f0-8198-6886eb863983)

**Comment by edigaryev** [comment]
Ubuntu 22.04.4 LTS already comes with this VirtIO sound module, however, it doesn't really work and the sound levels are jumping around a specific constant point:

<img width="1062" alt="Screenshot 2024-04-24 at 22 42 08" src="https://github.com/cirruslabs/tart/assets/85709/8106d5e3-a712-49ac-8f73-c5cb7761aaf6">

I'm not sure where you've obtained the 23.10 for `arm64`, but you can verify that you have the module either by going to `Settings → Sound` or by running the following commands:

```
$ grep CONFIG_SND_VIRTIO /boot/config-$(uname -r)
CONFIG_SND_VIRTIO=m
```

```
$ lsmod | grep '^virtio_snd'
virtio_snd             36864  4
```

**Comment by adit-s** [comment]
Hello @edigaryev. Thanks for getting back to me. 
I downloaded Ubuntu 23.10 for arm from [here](https://ubuntu.com/download/server/arm).
I gave up trying with 23.10 to stick with 22.04 LTS.
In trying to find a solution, it does look like the virtio package is included in the 5.15.0-105-generic (22.04 LTS) kernel.

Running the same commands on my installation of 22.04 gives the following results:
```
$ grep CONFIG_SND_VIRTIO /boot/config-$(uname -r)
CONFIG_SND_VIRTIO=m
```

```
$ lsmod | grep '^virtio_snd'
virtio_snd             40960  2
```

Iny my case, there isn't any sound.
I'm not sure, but it seems like it may be an issue with how the sound card is configured?
I don't really have any other ideas on how to get the sound working. If you do, I'd appreciate it if you can suggest them.

Thanks,
Adit

**Comment by edigaryev** [comment]
Something like this should work:

```shell
tart create --linux fedora
wget https://download.fedoraproject.org/pub/fedora/linux/releases/40/Workstation/aarch64/images/Fedora-Workstation-40-1.14.aarch64.raw.xz
unxz Fedora-Workstation-40-1.14.aarch64.raw
tart run --disk Fedora-Workstation-40-1.14.aarch64.raw fedora
```

The same approach for Manjaro.

**Comment by fkorotkov** [comment]
Seems like the issue is not with Tart but rather with particular distributions of Linux and support of virtio audio devices. Closing the issue.

**Comment by adit-s** [comment]
I agree it's not an issue with tart, but it would be nice if there was a way to make it work.
I'm sure it would help quite a few others as well.

Thanks for your help and guidance.

**Comment by adit-s** [comment]
I found a working solution. It's listed here for anyone else who comes across this post. 
Note: It only worked with Ubuntu 23.10 for me. 22.04 and 24.04 did not work.

I installed the Ubuntu 23.10 arm server image following the standard tart installation procedure.
Then installed the KDE desktop environment (sudo apt install kde-standard).
In the desktop, installed pavucontrol (sudo apt install pavucontrol).
Start the Pulse Audio Volume Control app from the desktop and under the configuration tab choose the 'Pro Audio' profile.

The sound works well for audio streams and video. It's not choppy or breaking like in some of the other setups.

I hope this helps.

**Comment by marcomastri** [comment]
I tried the suggestion on [this](https://bbs.archlinux.org/viewtopic.php?pid=2063665#p2063665) forum post (with a Fedora VM). For a while I thought the problem was gone, but playing a video on youtube still gives stuttering after some minutes of playback. Still, maybe it’s a step in the right direction.

**Comment by mightysashiman** [comment]
installed tart on my mac
installed and ran ubuntu:latest
upgraded to 22.04.1 LTS
installed ubuntu-desktop
now at last in the window managed environment, sound is not working.
Sound output device only shows up with "Dummy output".

**Comment by edigaryev** [comment]
@mightysashiman please see https://github.com/cirruslabs/tart/issues/795#issuecomment-2081302329, 22.04 won't likely work, while 23.10 could.
